src/core/providers/gdal/qgsgdalprovider.cpp

```diff
@@ -2454,6 +2454,172 @@ QString QgsGdalProviderMetadata::encodeUri( const QVariantMap &parts ) const
   return  QgsGdalProviderBase::encodeGdalUri( parts );
 }
 
+
+static bool _parseGpkgColons( const QString &src, QString &filename, QString &tablename )
+{
+  // GDAL accepts the following input format:  GPKG:filename:table
```

Comment:
Seeing all this logic in QGIS makes me think that ideally, it should't belong to QGIS, but GDAL should have an API to ask drivers to do that job.

Reply:
Tracked in https://github.com/OSGeo/gdal/issues/7261

Reply:
That would be quite nice if GDAL had such API... as you can see in QGIS code we're only handling few (most common?) cases